core/src/main/java/com/linecorp/armeria/common/metric/ServiceNaming.java

```diff
@@ -0,0 +1,150 @@
+ *
+ * @see RequestOnlyLog#serviceName()
+ * @see HttpService
+ * @see RpcService
+ */
+@FunctionalInterface
+public interface ServiceNaming {
+
+    /**
+     * Returns the {@link ServiceNaming} that always returns the given hard-coded service name.
+     */
+    static ServiceNaming of(String defaultServiceName) {
+        requireNonNull(defaultServiceName, "defaultServiceName");
+        return ctx -> defaultServiceName;
+    }
+
+    /**
+     * Returns the {@link ServiceNaming} that returns the simple name of innermost class from the given service.
+     * It is supposed to have a class name without a period. e.g. {@code HelloService}.
+     *
+     * @see Class#getSimpleName()
+     */
+    static ServiceNaming simpleTypeName() {
+        return ctx -> {
+            final RpcRequest rpcReq = ctx.rpcRequest();
+            if (rpcReq != null) {
+                final String serviceType = rpcReq.serviceType().getName();
+                if ("com.linecorp.armeria.internal.common.grpc.GrpcLogUtil".equals(serviceType)) {
+                    // Parse gRPC serviceName and methodName
+                    final String fullMethodName = rpcReq.method();
+                    final String fullServiceName = fullMethodName.substring(0, fullMethodName.lastIndexOf('/'));
+                    return fullServiceName.substring(fullServiceName.lastIndexOf('.') + 1);
+                } else {
+                    return serviceType;
+                }
+            }
+
+            Unwrappable unwrappable = ctx.config().service();
+            for (;;) {
+                final Unwrappable delegate = unwrappable.unwrap();
+                if (delegate != unwrappable) {
+                    unwrappable = delegate;
+                    continue;
+                }
+                final String name = delegate.getClass().getName();
+                return name.substring(name.lastIndexOf('.') + 1);
+            }
+        };
+    }
+
+    /**
+     * Returns the {@link ServiceNaming} that returns the full name of innermost class from the given service.
+     * It will be fully qualified name including the package name separated by a period.
+     * e.g. {@code com.foo.HelloService}.
+     *
+     * @see Class#getName()
+     */
+    static ServiceNaming fullTypeName() {
+        return ctx -> {
+            final RpcRequest rpcReq = ctx.rpcRequest();
+            if (rpcReq != null) {
+                final String serviceType = rpcReq.serviceType().getName();
+                if ("com.linecorp.armeria.internal.common.grpc.GrpcLogUtil".equals(serviceType)) {
+                    // Parse gRPC serviceName and methodName
+                    final String fullMethodName = rpcReq.method();
+                    final int methodIndex = fullMethodName.lastIndexOf('/');
+                    return fullMethodName.substring(0, methodIndex);
+                } else {
+                    return serviceType;
```

Comment:
Now we can remove the same code in `DefaultRequestLog`?
https://github.com/line/armeria/blob/aab237366debe0c0668c08688bd65deae9fe2424/core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java#L1054-L1068

Reply:
As you said at https://github.com/line/armeria/pull/3366#discussion_r604038359, currently `ServiceNaming.convert()` allow returning null. So, I kept duplicated codes in `DefaultRequestLog` to ensure same behavior when null is returned from `ServiceNaming.convert()`. This is quite possible from not only `ctx -> null`, but also user's mistake.
Plus, because of `ClientRequestContext`, most of the codes should be retained.

Reply:
> currently ServiceNaming.convert() allow returning null. S

If so, could add `@Nullable` annotation for the return type of `ServiceNaming.convert()`.

> I kept duplicated codes in DefaultRequestLog to ensure same behavior when null is returned from ServiceNaming.convert().

In that case, we can call the default service naming(`ServiceNaming.fullTypeName()`) as a fallback function for an RPC request or a service request.

```java
newServiceName = config.defaultServiceNaming().convert(sctx);
if (newServiceName == null && (ctx.rpcReq != null || config != null)) {
    newServiceName = ServiceNameing.fullTypeName();
}
```

Reply:
> could add @Nullable annotation for the return type of ServiceNaming.convert().

Sure. Updated. Also I revised `ServiceNaming` javadoc to describe a behavior of previous `defaultServiceName` as it is delegated to `ServiceNaming` now.

> as a fallback function

`DefaultRequestLog` is updated to call `ServiceNaming.fullTypeName()` as a fallback function and remove `getInnerMost` by deduplicating. Please check again.

Reply:
> nit. Even though I ran :*:javadoc successfully at local, it failed in appveyor 🤔

Ah, could you merge the master branch? Now, Javadoc is built with Java 15. (see #3426)

>  remove getInnerMostServiceName() as a deduplication

Could we also remove the RPC service name block use `ServiceNaming.fullTypeName()` together?

Reply:
> could you merge the master branch? Now, Javadoc is built with Java 15.

Oh thanks. I missed that PR.

> Could we remove the RPC service name block and use `ServiceNaming.fullTypeName()` also?

I think we cannot. As `ServiceNaming.convert()` needs `ServiceRequestContext`, it is not able to cover `ClientRequestContext`. That is, RPC service name block is needed for RPC client stub. e.g. https://github.com/line/armeria/blob/armeria-1.5.0/grpc/src/test/java/com/linecorp/armeria/common/grpc/GrpcMeterIdPrefixFunctionTest.java#L204.

Reply:
> As ServiceNaming.convert() needs ServiceRequestContext, it is not able to cover ClientRequestContext. 

Ah... it takes `ServiceRequestContext`. I think we can miss updating one of them if the same logic is placed in different locations.
How about adding `String ServiceNamingUtil.fullTypeName(RequestContext ctx)` to `internal.common` and delegate `ServiceNaming.fullTypeName()` to it?

By the way, I realized that the package of `ServiceNaming` is `armeria.common.metrics`. Don't we need to put `ServiceNaming` to `armeria.common`? Because it is also used for logging.

Reply:
> it is not able to cover ClientRequestContext

What if a user wants to customize service naming for clients? Don't we need to design `ServiceNaming` to use on the client-side?

How about adding `ServiceNaming` to `ClientOptions` in the following PR?

Reply:
> put `ServiceNaming` to `armeria.common`

SGTM

> How about adding ServiceNaming to ClientOptions in the following PR?

SGTM. Is it okay for me to write an issue in advance?

> How about adding `String ServiceNamingUtil.fullTypeName(RequestContext ctx)` to `internal.common` and delegate `ServiceNaming.fullTypeName()` to it?

If putting all into one util function, `ServiceNamingUtil.fullTypeName()`, it seems that the responsibilities for `ServiceRequestContext` and `ClientRequestContext` are mixed and the readability will be worse. However, I agreed not to keep the same logic at several points for a maintenance. As SoC(Separation of Concerns), how about introducing an util function just for RPC service name part?

Reply:
> Is it okay for me to write an issue in advance?

You can do whatever you are comfortable with.


>  String ServiceNamingUtil.fullTypeRpcServiceName(RpcRequest)?

Sounds good.